docs/Configuration.md
# Configuration

By default, Vue styleguidist will look for `styleguide.config.js` file in your project’s root folder. You can change the location of the config file using `--config` [CLI](/docs/CLI.md) option.

## `assetsDir`

Type: `String`, optional

Your application static assets folder, will be accessible as `/` in the style guide dev server.

## `codeSplit`

Type: `Boolean`, default: true

By default vue-styleguidist will build one single bundle for all your javascript. When this flag is on, the editor (CodeMirror) is loaded as a separate bundle and so is the compiler. Each of those 2 bundles are about 400Kb. It allows for a faster initial load.

> **Note**: When you use this option and have a babel-loader in your webpack config, make sure that you either ignore `node_modules` in your babel loader or that you have the [proper plugin](https://babeljs.io/docs/en/babel-plugin-syntax-dynamic-import) for code splitting with babel. If not, you might get the following error:
>
> `SyntaxError: Support for the experimental syntax 'dynamicImport' isn't currently enabled`

## `compilerPackage`

Type: `String`, optional

Path or name of a module that exports a `compile` and a `getImports` function. It will be used to compile your live examples. vue-styleguidist will do `require(config.compilerPackage)` to get the `compile` function. 
- Checkout [examples-loader.ts](https://github.com/vue-styleguidist/vue-styleguidist/blob/dev/packages/vue-styleguidist/src/loaders/examples-loader.ts) to see the implementation on the node side. 
- See [Preview.js](https://github.com/vue-styleguidist/vue-styleguidist/blob/dev/packages/vue-styleguidist/src/client/rsg-components/Preview/Preview.js) to see how it is used in the frontend.

## `compilerConfig`

Type: `Object`, default: `{ objectAssign: 'Object.assign' }`

Styleguidist uses `vue-inbrowser-compiler` to bundle typescript/ES6 modules code on the frontend. These options are passed to the compile function.

## `components`

Type: `String`, `Function` or `Array`, default: `src/components/**/*.vue`

- when `String`: a [glob pattern](https://github.com/isaacs/node-glob#glob-primer) that matches all your component modules.
- when `Function`: a function that returns an array of module paths.
- when `Array`: an array of module paths.

All paths are relative to config folder.

See examples in the [Components section](/docs/Components.md#components).

## `context`

Type: `Object`, optional

Modules that will be available for examples. You can use it for utility functions like Lodash or for data fixtures.

```javascript
module.exports = {
  context: {
    map: 'lodash/map',
    users: path.resolve(__dirname, 'fixtures/users')
  }
}
```

Then you can use them in any example:

```jsx
<Message>{map(users, 'name').join(', ')}</Message>
```

## `contextDependencies`

Type: `String[]`, optional

Array of absolute paths that allow you to specify absolute paths of directories to watch for additions or removals of components.

By default Styleguidist uses common parent directory of your components.

```javascript
module.exports = {
  contextDependencies: [path.resolve(__dirname, 'lib/components')]
}
```

## `configureServer`

Type: `Function`, optional

Function that allows you to add endpoints to the underlying Express server:

```javascript
module.exports = {
  configureServer(app) {
    // `app` is the instance of the express server running Styleguidist
    app.get('/custom-endpoint', (req, res) => {
      res.status(200).send({ response: 'Server invoked' })
    })
  }
}
```

Your components will be able to invoke the URL `http://localhost:6060/custom-endpoint` from their examples.

## `copyCodeButton`

Type: `Boolean`, default: `false`

Adds a little button on the top right hand corner of the editor to copy the contents of the editor into the clipboard.

## `dangerouslyUpdateWebpackConfig`

Type: `Function`, optional

> **Warning:** You are likely to break Vue styleguidist using this options, try to use [webpackConfig](#webpackconfig) option instead.

Allows you to modify webpack config without any restrictions.

```javascript
module.exports = {
  dangerouslyUpdateWebpackConfig(webpackConfig, env) {
    // WARNING: inspect Vue styleguidist Webpack config before modifying it, otherwise you may break Vue styleguidist
    console.log(webpackConfig)
    webpackConfig.externals = {
      jquery: 'jQuery'
    }
    return webpackConfig
  }
}
```

## `defaultExample`

Type: `Boolean` or `String`, default: `false`

For components that do not have an example, a default one can be used. When set to `true`, the [DefaultExample.md](https://github.com/vue-styleguidist/vue-styleguidist/blob/delivery/packages/vue-styleguidist/scripts/templates/DefaultExample.md) is used, or you can provide the path to your own example Markdown file.

When writing your own default example file, `__COMPONENT__` will be replaced by the actual component name at compile time.

## `displayOrigins`

Type: `Boolean`, default: `false`

When turned on, displays if a prop, event, slot or methods is from the current file or is configured in a mixin or an extended component.

If it is external, it displays the name of the component and on hover displays the relative path to the file.

## `getComponentPathLine`

Type: `Function`, default: component file name

Function that returns a component path line (displayed under the component name).

For example, instead of `components/Button/Button.vue` you can print `import Button from 'components/Button';`:

```javascript
const path = require('path')
module.exports = {
  getComponentPathLine(componentPath) {
    const name = path.basename(componentPath, '.vue')
    const dir = path.dirname(componentPath)
    return `import ${name} from '${dir}';`
  }
}
```

## `editorConfig`

Type: `Object`, default: [scripts/schemas/config.js](https://github.com/vue-styleguidist/vue-styleguidist/blob/delivery/packages/vue-styleguidist/src/scripts/schemas/config.ts#L103-L112)

Source code editor options, see [CodeMirror docs](https://codemirror.net/doc/manual.html#config) for all available options.

> **NOTE :** From version 4.0.0, Prism is the editor by default as it is much lighter. Turn off [simpleEditor](#simpleEditor) to use CodeMirror and leverage this config.
>
> ```js
> module.exports = {
>   // ...
>   editorConfig: {
>     theme: 'base16-light'
>   },
>   simpleEditor: false
> }
> ```

## `getExampleFilename`

Type: `Function`, default: finds `Readme.md` or `ComponentName.md` in the component folder

Function that returns examples file path for a given component path.

For example, instead of `Readme.md` you can use `ComponentName.examples.md`:

```javascript
module.exports = {
  getExampleFilename(componentPath) {
    return componentPath.replace(/\.jsx?$/, '.examples.md')
  }
}
```

## `enhancePreviewApp`

Type `String`, optional

Allows to register **global plugins** and **global directives** in the examples preview or **Vue 3** style guides.

For example:

Reference the preview file path in the `styleguide.config.js`

```js
// styleguide.config.js
export default {
  // ...
  enhancePreviewApp: path.resolve(__dirname, 'styleguide/preview.js')
}
```

Then update your app in it with whatever you want.

```js
// styleguide/preview.js
import { defineEnhanceApp } from 'vue-styleguidist/helpers'
import focusDirective from '../src/directives/v-focus';

// The export here MUST be default or module.export
// this is what is imported by the styleguide
export default defineEnhanceApp((app) => {
  app.directive('focus', focusDirective)
})
```

> NOTE: If you use TypeScript, this file can be TypeScript as well

## `exampleMode`

Type: `'collapse' | 'hide' | 'expand'`, default: `collapse`

Defines the initial state of the example code tab:

- `collapse`: collapses the tab by default.
- `hide`: hide the tab and it can´t be toggled in the UI.
- `expand`: expand the tab by default.

## `ignore`

Type: `String[]`, default: `['**/__tests__/**']`

Array of [glob pattern](https://github.com/isaacs/node-glob#glob-primer) that should not be included in the style guide.

> **Note:** You should pass glob patterns, for example, use `**/components/Button.vue` instead of `components/Button.vue`.

## `jssThemedEditor`

Type: `Boolean`, default: `true`

Should integrated PrismJs editors be themed using JSS in the theme option.

If you want to use a theme defined in CSS, set this to false and require the CSS file in the `require` config.

[prism themes repo](https://github.com/PrismJS/prism-themes/)

> **Note:** There is [a bug in prism editor](https://github.com/satya164/react-simple-code-editor/issues/56) that prevents us from changing the editor's background and font-color easily. Workaround: If you get a theme with
>
> ```css
> pre[class*='language-'] {
>   padding: 1em;
>   margin: 0.5em 0;
>   overflow: auto;
>   background: #1e1e1e;
> }
> ```
>
> Use the `div.prism-editor` class to change the background instead
>
> ```css
> pre[class*='language-'] {
>   padding: 1em;
>   margin: 0.5em 0;
>   overflow: auto;
> }
>
> div.prism-editor {
>   background: #1e1e1e;
> }
> ```
>
> checkout [examples/customised](https://github.com/vue-styleguidist/vue-styleguidist/blob/dev/examples/customised/styleguide/vsc-prism.css) for an example of implementation

## `jsxInComponents`

Type: `Boolean`, default: `true`

Do your components contain JSX syntax? Since some TypeScript syntax can sometimes conflict with JSX, it ca be useful to disble it when needed. The following TypeScript code would fail parsing if this flag is not set to false.

```typescript
function initDatepicker() {
  ;(<any>window).$.datetimepicker({
    // babel parser will think that `<any>` is actually jsx
    //...
  })
}
```

## `jsxInExamples`

Type: `Boolean`, default: `false`

If your examples are written in JSX, you will need this flag on.

```jsx
export default {
  render() {
    return <Button />
  }
}
```

> **Note:** When this flag is on, the pseudo-jsx examples will not work anymore. Example should be written with proper vue syntax.

## `logger`

Type: `Object`, by default will use `console.*` in CLI or nothing in Node.js API

Custom logger functions:

```javascript
module.exports = {
  logger: {
    // One of: info, debug, warn
    // Suppress messages
    info: () => {},
    // Override display function
    warn: message => console.warn(`NOOOOOO: ${message}`)
  }
}
```

## `locallyRegisterComponents`

Type: `Boolean`, default: `false`

By default, `vue-styleguidist` registers all components globally. This can be an issue when:

- Multiple components are sharing the same name OR
- Components are changing behaviour if another component is registered

In this case, set `locallyRegisterComponents` to `true`. It will register components only in the examples of their documentation.

Though if you need to register an additionl component and you are forced to use this behaviour, proceed like this:

1.  Write examples using SFC format
1.  Explicitly require or import all needed components
1.  Register them in your example

```vue
<script>
import MyButton from './src/components/MyButton.vue'
import MyIcon from './src/components/MyIcon.vue'

export default {
  components: { MyButton, MyIcon }
}
</script>

<template>
  <MyButton><MyIcon name="Save" />Save Form</MyButton>
</template>
```

> **Note** This can be done as well using a `new Vue()` script

## `minimize`

Type: `boolean`, default: `true`

If you wish to remove minimization from the build process to help with debugging or to accelate netlify build time on PR, turn this one off.

## `mountPointId`

Type: `string`, default: `rsg-root`

The ID of a DOM element where Styleguidist mounts.

## `pagePerSection`

Type: `Boolean`, default: `false`

Render one section or component per page.

If `true`, each section will be a single page.

The value may depends on a current environment:

```javascript
module.exports = {
  pagePerSection: process.env.NODE_ENV !== 'production'
}
```

To isolate section’s children as single pages (subroutes), add `sectionDepth` into each section with the number of subroutes (depth) to render as single pages.

For example:

```javascript
module.exports = {
  pagePerSection: true,
  sections: [
    {
      name: 'Documentation',
      sections: [
        {
          name: 'Files',
          sections: [
            {
              name: 'First File'
            },
            {
              name: 'Second File'
            }
          ]
        }
      ],
      // Will show "Documentation" and "Files" as single pages, filtering its children
      sectionDepth: 2
    },
    {
      name: 'Components',
      sections: [
        {
          name: 'Buttons',
          sections: [
            {
              name: 'WrapperButton'
            }
          ]
        }
      ]
      // Will show "Components" as single page, filtering its children
      sectionDepth: 1,
    },
    {
      name: 'Examples',
      sections: [
        {
          name: 'Case 1',
          sections: [
            {
              name: 'Buttons'
            }
          ]
        }
      ]
      // There is no subroutes, "Examples" will show all its children on a page
      sectionDepth: 0,
    }
  ]
}
```

## `printBuildInstructions`

Type: `Function`, optional

Function that allows you to override the printing of build messages to console.log.

```javascript
module.exports = {
  printBuildInstructions(config) {
    console.log(
      `Style guide published to ${config.styleguideDir}. Something else interesting.`
    )
  }
}
```

## `printServerInstructions`

Type: `Function`, optional

Function that allows you to override the printing of local dev server messages to console.log.

```javascript
module.exports = {
  serverHost: 'your-domain',
  printServerInstructions(config, { isHttps }) {
    console.log(`Local style guide: http://${config.serverHost}`)
  }
}
```

## `previewDelay`

Type: `Number`, default: 500

Debounce time in milliseconds used before render the changes from the editor. While typing code the preview will not be updated.

## `propsParser`

Type: `Function`, optional

Function that allows you to override the mechanism used to parse props from a source file. Default mechanism is using [vue-docgen-api](https://github.com/vue-styleguidist/vue-docgen-api) to parse props.

```javascript
module.exports = {
  propsParser(filePath, source) {
    return require('vue-docgen-api').parse(filePath)
  }
}
```

## `progressBar`

Type: `Boolean`, default: true

Should styleguidist show a progress bar using `webpack.ProgressPlugin` while building.

> **Note:**: Builds on CI might be slower when this is active. You might want to make this parameter dependent on CI environement variables.

## `require`

Type: `String[]`, optional

Modules that are required for your style guide. Useful for third-party styles or polyfills.

```javascript
module.exports = {
  require: [
    'babel-polyfill',
    path.join(__dirname, 'styleguide/styles.css')
  ]
}
```

> **Note:** This will add a separate webpack entry for each array item.

Don’t forget to add webpack loaders for each file you add here. For example, to require a CSS file you’ll need:

```javascript
module.exports = {
  webpackConfig: {
    module: {
      rules: [
        {
          test: /\.css$/,
          use: ['style-loader', 'css-loader']
        }
      ]
    }
  }
}
```

See [Configuring webpack](/docs/Webpack.md) for mode details.

## `renderRootJsx`

Type: `String`, optional

Modifies the root component of the previews of the examples, receives as a parameter the preview component, must return a [component jsx](https://vuejs.org/v2/guide/render-function.html). It is useful when you want to create a container to all your examples by default, especially when you use a third library.

> NOTE: if you are using styleguidist with Vue 3, you should use the [enhancePreviewApp](#enhancePreviewApp) option instead of `renderRootJsx` because the latter is not compatible with Vue 3.

An example

```javascript
// config/styleguide.root.js
export default function (previewComponent) {
  return {
    render(createElement) {
      return createElement(previewComponent);
    }
  },
}
```

```javascript
module.exports = {
  renderRootJsx: path.join(__dirname, 'config/styleguide.root.js')
}
```

## `ribbon`

Type: `Object`, optional

Shows 'Fork Me' ribbon in the top-right corner. If `ribbon` key is present, then it's required to add `url` property; `text` property is optional. If you want to change styling of the ribbon, please, refer to the [theme section](#theme).

```javascript
module.exports = {
  ribbon: {
    url: 'http://example.com/',
    text: 'Fork me on GitHub'
  }
}
```

## `sections`

Type: `Array`, optional

Allows components to be grouped into sections with a title and overview content. Sections can also be content only, with no associated components (for example, a textual introduction). Sections can be nested.

See examples of [sections configuration](/docs/Components.md#sections).

## `serverHost`

Type: `String`, default: `0.0.0.0`

Dev server name.

## `serverPort`

Type: `Number`, default: `6060`

Dev server port.

## `showSidebar`

Type: `Boolean`, default: `true`

Toggle sidebar visibility. Sidebar will be hidden when opening components or examples in isolation mode even if this value is set to `true`. When set to `false`, sidebar will always be hidden.

## `simpleEditor`

Type: `Boolean`, default: `false`

Avoid loading CodeMirror and reduce bundle size significantly, use [prism.js](https://prismjs.com/) for code highlighting. Warning: editor options will not be mapped over.

## `skipComponentsWithoutExample`

Type: `Boolean`, default: `false`

Ignore components that don’t have an example file (as determined by [getExampleFilename](#getexamplefilename)). These components won’t be accessible from other examples unless you [manually `require` them](/docs/Cookbook.md#how-to-hide-some-components-in-style-guide-but-make-them-available-in-examples).

## `styleguideComponents`

Type: `Object`, optional

Override React components used to render the style guide.

```javascript
module.exports = {
  styleguideComponents: {
    Wrapper: path.join(__dirname, 'styleguide/components/Wrapper'),
    StyleGuideRenderer: path.join(
      __dirname,
      'styleguide/components/StyleGuide'
    )
  }
}
```

See an example of [customized style guide](https://github.com/vue-styleguidist/vue-styleguidist/tree/delivery/examples/customised).

If you want to wrap, rather than replace a component, make sure to import the default implementation using the full path to `vue-styleguidist`. See an example of [wrapping a Styleguidist component](https://github.com/vue-styleguidist/vue-styleguidist/blob/delivery/examples/customised/styleguide/components/SectionsRenderer.js).

**Note**: these components are not guaranteed to be safe from breaking changes in vue-styleguidist updates.

## `styleguideDir`

Type: `String`, default: `styleguide`

Folder for static HTML style guide generated with `styleguidist build` command.

## `styleguidePublicPath`

Type: `String`, optional

Configures the prefix of the build and server urls.

Example: If set to `/mystyleguide` the url for dev server will be `htp://localhost:6060/mystyleguide`

## `styles`

Type: `Object`, `String` or `Function`, optional

Customize styles of any Styleguidist’s component using an object, a function returning said object or a file path to a file exporting said styles.

See example in the [cookbook](/docs/Cookbook.md#how-to-change-styles-of-a-style-guide).

> **Note:** Using a function allows access to theme variables as seen in the example below. See available [theme variables](https://github.com/styleguidist/react-styleguidist/blob/master/src/client/styles/theme.ts). The returned object folows the same format as when configured as a litteral.

```javascript
module.exports = {
  styles: function (theme) {
    return {
      Logo: {
        logo: {
          // we can now change the color used in the logo item to use the theme's `link` color
          color: theme.color.link
        }
      }
    }
  }
}
```

**Note:** If using a file path, it has to be absolute or relative to the config file.

## `template`

Type: `Object` or `Function`, optional.

Change HTML for the style guide app.

An object with options to add a favicon, meta tags, inline JavaScript or CSS, etc. See [@vxna/mini-html-webpack-template docs](https://www.npmjs.com/package/@vxna/mini-html-webpack-template).

```javascript
module.exports = {
  template: {
    favicon: 'https://assets-cdn.github.com/favicon.ico'
  }
}
```

A function that returns an HTML string, see [mini-html-webpack-plugin docs](https://github.com/styleguidist/mini-html-webpack-plugin#custom-templates).

## `theme`

Type: `Object` or `String`, optional

Customize style guide UI fonts, colors, etc. using a theme object or the path to a file exporting such object.

The path is relative to the config file or absolute.

See example in the [cookbook](/docs/Cookbook.md#how-to-change-styles-of-a-style-guide).

> **Note:** See available [theme variables](https://github.com/styleguidist/react-styleguidist/blob/master/src/client/styles/theme.ts).
>
> **Note:** Styles use [JSS](https://github.com/cssinjs/jss/blob/master/docs/jss-syntax.md) with these plugins: [jss-plugin-isolate](https://github.com/cssinjs/jss/tree/master/packages/jss-plugin-isolate), [jss-plugin-nested](https://github.com/cssinjs/jss/tree/master/packages/jss-plugin-nested), [jss-plugin-camel-case](https://github.com/cssinjs/jss/tree/master/packages/jss-plugin-camel-case), [jss-plugin-default-unit](https://github.com/cssinjs/jss/tree/master/packages/jss-plugin-default-unit), [jss-plugin-compose](https://github.com/cssinjs/jss/tree/master/packages/jss-plugin-compose) and [jss-plugin-global](https://github.com/cssinjs/jss/tree/master/packages/jss-plugin-global).
>
> **Note:** Use [React Developer Tools](https://github.com/facebook/react) to find component and style names. For example a component `<LogoRenderer><h1 className="rsg--logo-53">` corresponds to an example above.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [4.45.0](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.29...v4.45.0) (2022-04-29)


### Features

* allow use of multiline examples and types ([a95a306](https://github.com/vue-styleguidist/vue-styleguidist/commit/a95a306a67fbf2bbdf0e7f6bddba143192ffcfa1)), closes [#1166](https://github.com/vue-styleguidist/vue-styleguidist/issues/1166)





## [4.44.29](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.28...v4.44.29) (2022-04-18)

**Note:** Version bump only for package vue-styleguidist





## [4.44.28](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.27...v4.44.28) (2022-04-11)


### Bug Fixes

* 🐛 webpack5 serve undefined process ([54c7e1a](https://github.com/vue-styleguidist/vue-styleguidist/commit/54c7e1a4983a7b1253000f9ca55fd2f9ef6a64c3)), closes [#1074](https://github.com/vue-styleguidist/vue-styleguidist/issues/1074)





## [4.44.27](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.25...v4.44.27) (2022-03-26)


### Bug Fixes

* add missing dependency back ([f6a8944](https://github.com/vue-styleguidist/vue-styleguidist/commit/f6a8944356cacd699db7793a5433f2c6b8849ae8))
* check if the current release is already deployed ([4d3164e](https://github.com/vue-styleguidist/vue-styleguidist/commit/4d3164ead0f1750ebea416253f2ad93d09ef1369))
* make sure the proper tag is created ([acc88e6](https://github.com/vue-styleguidist/vue-styleguidist/commit/acc88e6c8571c27f9e2a70632055e4a3d5d8bfc0))
* **plugin:** use webpack version to only remove when necessary ([cdd8c84](https://github.com/vue-styleguidist/vue-styleguidist/commit/cdd8c841d2ec69439cf2488e673b863bb2ec20ae))
* try creating a release using lerna version ([db9de67](https://github.com/vue-styleguidist/vue-styleguidist/commit/db9de67c14ab6d33f4f109be65a5e453ae265a8f))
* use continue-on-error ([5884599](https://github.com/vue-styleguidist/vue-styleguidist/commit/5884599a51cc3143cb6bc1b7bedb0f10596c37a3))





## [4.44.26](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.25...v4.44.26) (2022-03-25)


### Bug Fixes

* add missing dependency back ([f6a8944](https://github.com/vue-styleguidist/vue-styleguidist/commit/f6a8944356cacd699db7793a5433f2c6b8849ae8))
* **plugin:** use webpack version to only remove when necessary ([cdd8c84](https://github.com/vue-styleguidist/vue-styleguidist/commit/cdd8c841d2ec69439cf2488e673b863bb2ec20ae))





## [4.44.25](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.24...v4.44.25) (2022-03-25)


### Bug Fixes

* allow both wp 4 and wp5 to work with plugin ([e6a566b](https://github.com/vue-styleguidist/vue-styleguidist/commit/e6a566b0b09bf7de2bf0b9bf49a62c9052df86a6))
* make cli plugin work with wp 5 ([47dce60](https://github.com/vue-styleguidist/vue-styleguidist/commit/47dce605552cefa1d1c0739cc34b38aa5231510e))





## [4.44.24](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.23...v4.44.24) (2022-03-25)


### Bug Fixes

* filter html plugin in vue cli plugin ([a2183d2](https://github.com/vue-styleguidist/vue-styleguidist/commit/a2183d296ff9d5d1a22a43b43f4638ff3af710ee))





## [4.44.23](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.22...v4.44.23) (2022-03-25)


### Bug Fixes

* last babel cheks for const initialization ([8587549](https://github.com/vue-styleguidist/vue-styleguidist/commit/85875493a550f0f9f6f61262d91ba1013b88e796)), closes [#1299](https://github.com/vue-styleguidist/vue-styleguidist/issues/1299)
* make sure old version of node won't fail ([d6a423c](https://github.com/vue-styleguidist/vue-styleguidist/commit/d6a423c90e25dd9e40dc7db09009962b4cd21e50)), closes [#1300](https://github.com/vue-styleguidist/vue-styleguidist/issues/1300)





## [4.44.22](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.21...v4.44.22) (2022-03-18)


### Bug Fixes

* catch error if vue is not found ([b2f9a5c](https://github.com/vue-styleguidist/vue-styleguidist/commit/b2f9a5ccb34768730e5a4ca423904de3457bec62)), closes [#1295](https://github.com/vue-styleguidist/vue-styleguidist/issues/1295)
* require version instead of pkg ([da0f721](https://github.com/vue-styleguidist/vue-styleguidist/commit/da0f7218f097e45e6857974234d1a61d68b144a2))





## [4.44.21](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.20...v4.44.21) (2022-03-18)


### Bug Fixes

* publish proper postinstall ([5ad78fc](https://github.com/vue-styleguidist/vue-styleguidist/commit/5ad78fcfd645aa0375035656b2050bf61a733e09))





## [4.44.20](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.19...v4.44.20) (2022-03-18)


### Bug Fixes

* create postinstall to adapt vue3 ([7de0803](https://github.com/vue-styleguidist/vue-styleguidist/commit/7de0803c670a0532af409f6798cc008ffb2e13bf))
* fix postinstall script ([1dd108a](https://github.com/vue-styleguidist/vue-styleguidist/commit/1dd108a5031993755c24d0079e4281656904987d))





## [4.44.19](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.18...v4.44.19) (2022-03-16)


### Bug Fixes

* update esm import of html plugin ([42bfc53](https://github.com/vue-styleguidist/vue-styleguidist/commit/42bfc53f6f08b09ac792243ace3a3abfb162cc7a))





## [4.44.18](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.17...v4.44.18) (2022-03-10)


### Bug Fixes

* 🐛 Allow imported TS in JS SFCs ([ccc06ce](https://github.com/vue-styleguidist/vue-styleguidist/commit/ccc06ce6e256ab351f048f3e71182c84038a4e35))
* Indentation ([da0d6d9](https://github.com/vue-styleguidist/vue-styleguidist/commit/da0d6d91667b9fc56f39ea060891ed4b1ccbd04a))





## [4.44.17](https://github.com/vue-styleguidist/vue-styleguidist/compare/v4.44.16...v4.44.17) (2022-02-28)


### Bug Fixes


scripts/schemas/config.js
// If you want to access any of these options in React, don’t forget to update CLIENT_CONFIG_OPTIONS array
// in loaders/styleguide-loader.js

const DEFAULT_COMPONENTS_PATTERN = `src/{components,Components}/**/*.vue`;

const path = require('path');
const startCase = require('lodash/startCase');
const kleur = require('kleur');
const logger = require('glogg')('rsg');
const fileExistsCaseInsensitive = require('react-styleguidist/scripts/utils/findFileCaseInsensitive');
const getUserPackageJson = require('react-styleguidist/scripts/utils/getUserPackageJson');
const StyleguidistError = require('react-styleguidist/scripts/utils/error');
const findUserWebpackConfig = require('../utils/findUserWebpackConfig');
const consts = require('../consts');

module.exports = {
	assetsDir: {
		type: 'existing directory path',
		example: 'assets',
	},
	compilerConfig: {
		type: 'object',
		default: {
			objectAssign: 'Object.assign',
		},
	},
	// `components` is a shortcut for { sections: [{ components }] },
	// see `sections` below
	components: {
		type: ['string', 'function', 'array'],
		example: 'components/**/[A-Z]*.vue',
	},
	configDir: {
		process: (value, config, rootDir) => rootDir,
	},
	context: {
		type: 'object',
		default: {},
		example: {
			map: 'lodash/map',
		},
	},
	contextDependencies: {
		type: 'array',
	},
	configureServer: {
		type: 'function',
	},
	dangerouslyUpdateWebpackConfig: {
		type: 'function',
	},
	defaultExample: {
		type: ['boolean', 'existing file path'],
		default: false,
		process: val =>
			val === true ? path.resolve(__dirname, '../templates/DefaultExample.md') : val,
	},
	exampleMode: {
		type: 'string',
		process: (value, config) => {
			return config.showCode === undefined ? value : config.showCode ? 'expand' : 'collapse';
		},
		default: 'collapse',
	},
	getComponentPathLine: {
		type: 'function',
		default: componentPath => componentPath,
	},
	getExampleFilename: {
		type: 'function',
		default: componentPath => {
			const files = [
				path.join(path.dirname(componentPath), 'Readme.md'),
				componentPath.replace(path.extname(componentPath), '.md'),
			];

			for (const file of files) {
				const existingFile = fileExistsCaseInsensitive(file);
				if (existingFile) {
					return existingFile;
				}
			}

			return false;
		},
		example: componentPath => componentPath.replace(/\.jsx?$/, '.examples.md'),
	},
	ignore: {
		type: 'array',
		default: ['**/__tests__/**', `**/*.test.vue`, `**/*.spec.vue`, '**/*.d.ts'],
	},
	highlightTheme: {
		type: 'string',
		default: 'base16-light',
		deprecated: 'Use the theme property in the editorConfig option instead',
	},
	editorConfig: {
		type: 'object',
		process: (value, config) => {
			const defaults = {
				theme: 'base16-light',
				mode: 'jsx',
				lineWrapping: true,
				smartIndent: false,
				matchBrackets: true,
				viewportMargin: Infinity,
				lineNumbers: false,
			};
			return Object.assign(
				{},
				defaults,
				config.highlightTheme && {
					theme: config.highlightTheme,
				},
				value
			);
		},
	},
	navigation: {
		type: 'boolean',
		default: false,
		deprecated: 'Use pagePerSection option instead',
	},
	mixins: {
		type: 'array',
		default: [],
		example: ['path/to/mixin.js', 'path/to/created.js'],
		deprecated: 'Use renderRootJsx option instead',
	},
	logger: {
		type: 'object',
	},
	mountPointId: {
		type: 'string',
		default: 'rsg-root',
	},
	pagePerSection: {
		type: 'boolean',
		default: false,
	},
	previewDelay: {
		type: 'number',
		default: 500,
	},
	printBuildInstructions: {
		type: 'function',
	},
	printServerInstructions: {
		type: 'function',
	},
	propsParser: {
		type: 'function',
	},
	require: {
		type: 'array',
		default: [],
		example: ['babel-polyfill', 'path/to/styles.css'],
	},
	renderRootJsx: {
		type: 'directory path',
	},
	ribbon: {
		type: 'object',
		example: {
			url: 'http://example.com/',
			text: 'Fork me on GitHub',
		},
	},
	sections: {
		type: 'array',
		default: [],
		process: (val, config) => {
			if (!val) {
				// If root `components` isn't empty, make it a first section
				// If `components` and `sections` weren’t specified, use default pattern
				const components = config.components || DEFAULT_COMPONENTS_PATTERN;
				return [
					{
						components,
					},
				];
			}
			return val;
		},
		example: [
			{
				name: 'Documentation',
				content: 'Readme.md',
			},
			{
				name: 'Components',
				components: './lib/components/**/[A-Z]*.js',
			},
		],
	},
	serverHost: {
		type: 'string',
		default: '0.0.0.0',
	},
	serverPort: {
		type: ['number', 'string'],
		default: 6060,
	},
	showCode: {
		type: 'boolean',
		default: false,
		deprecated: 'Use exampleMode option instead',
	},
	showUsage: {
		type: 'boolean',
		default: false,
		deprecated: 'Use usageMode option instead',
	},
	showSidebar: {
		type: 'boolean',
		default: true,
	},
	skipComponentsWithoutExample: {
		type: 'boolean',
		default: false,
	},
	sortProps: {
		type: 'function',
	},
	styleguideComponents: {
		type: 'object',
	},
	styleguideDir: {
		type: 'directory path',
		default: 'styleguide',
	},
	styleguidePublicPath: {
		type: 'string',
		default: '',
	},
	styles: {
		type: 'object',
		default: {},
		example: {
			Logo: {
				logo: {
					fontStyle: 'italic',
				},
			},
		},
	},
	template: {
		type: ['object', 'function'],
		default: {},
		process: val => {
			if (typeof val === 'string') {
				throw new StyleguidistError(
					`${kleur.bold(
						'template'
					)} config option format has been changed, you need to update your config.`,
					'template'
				);
			}
			return val;
		},
	},
	theme: {
		type: 'object',
		default: {},
		example: {
			link: 'firebrick',
			linkHover: 'salmon',
		},
	},
	title: {
		type: 'string',
		process: val => {
			if (val) {
				return val;
			}
			const name = getUserPackageJson().name || '';
			return `${startCase(name)} Style Guide`;
		},
		example: 'My Style Guide',
	},
	updateDocs: {
		type: 'function',
	},
	updateExample: {
		type: 'function',
		default: props => {
			if (props.lang === 'example') {
				props.lang = 'js';
				logger.warn(
					'"example" code block language is deprecated. Use "js", "jsx" or "javascript" instead:\n' +
						consts.DOCS_DOCUMENTING
				);
			}
			return props;
		},
	},
	updateWebpackConfig: {
		type: 'function',
		removed: `Use "webpackConfig" option instead:\n${consts.DOCS_WEBPACK}`,
	},
	usageMode: {
		type: 'string',
		process: (value, config) => {
			return config.showUsage === undefined ? value : config.showUsage ? 'expand' : 'collapse';
		},
		default: 'collapse',
	},
	verbose: {
		type: 'boolean',
		default: false,
	},
	version: {
		type: 'string',
	},
	vuex: {
		type: 'directory path',
		deprecated: 'Use renderRootJsx option instead',
	},
	webpackConfig: {
		type: ['object', 'function'],
		process: val => {
			if (val) {
				return val;
			}

			const file = findUserWebpackConfig();
			if (file) {
				logger.info(`Loading webpack config from:\n${file}`);
				// eslint-disable-next-line import/no-dynamic-require
				return require(file);
			}

			logger.warn(
				'No webpack config found. ' +
					'You may need to specify "webpackConfig" option in your style guide config:\n' +
					consts.DOCS_WEBPACK
			);

			return undefined;
		},
		example: {
			module: {
				rules: [
					{
						test: /\.vue$/,
						exclude: /node_modules/,
						loader: 'vue-loader',
						options: {},
					},
					{
						test: /\.js$/,
						exclude: /node_modules/,
						loader: 'babel-loader',
					},
					{
						test: /\.css$/,
						loader: 'style-loader!css-loader',
					},

					{
						test: /\.json$/,
						loader: 'json-loader',
					},
					{
						exclude: [/\.html$/, /\.(js|jsx)$/, /\.css$/, /\.vue$/, /\.json$/],
						loader: 'url-loader',
						query: {
							limit: 10000,
							name: 'static/media/[name].[hash:8].[ext]',
						},
					},
				],
			},
		},
	},
};
